core/src/main/java/org/apache/calcite/rex/RexCall.java

```diff
@@ -59,6 +59,8 @@ protected RexCall(
     this.type = Objects.requireNonNull(type);
     this.op = Objects.requireNonNull(op);
     this.operands = ImmutableList.copyOf(operands);
+    this.digest = computeDigest(isA(SqlKind.CAST)
```

Comment:
I would rather avoid public access of `digest` field.

It might be the case that `RexCall` is created and its digest is not used. In that case digest computation would be a waste of resources.

Reply:
So we should make `digest` field to private and provide a method like `getDigest`.Do you think is it ok?

Reply:
private is not an option as it breaks backward compatibility.
For now, direct `.digest` should be replaced with `.toString()`

Reply:
if `digest` is not private,we cannot disallow using of  `digest` in other place.On the other handle ,the semantics of `.toString()` is not equal to `digest`.

Reply:
> if digest is not private,we cannot disallow using of digest in other place

That's right. However we shouldn't break backward compatibility with little reason.

>On the other hand ,the semantics of .toString() is not equal to digest.

In fact, `RexNode#toString` is used a lot in Calcite. For instance:

```
    // If the join is the same, we bail out
    if (newJoin instanceof Join) {
      final RexNode newCondition = ((Join) newJoin).getCondition();
      if (join.getCondition().toString().equals(newCondition.toString())) {
```

It might be worth refactoring, however let's fix `coalesce` bug first, then discuss `digest vs toString` refactor?

Reply:
Ok,I will fix this bug first and start another issue to discuss the refactor later.